File: exercism/python/atbash-cipher/atbash_cipher.py
from collections import defaultdict
from itertools import chain, zip_longest, repeat
from string import digits, ascii_lowercase

rev = ''.join(reversed(ascii_lowercase))
cipher = defaultdict(lambda: None, zip(ascii_lowercase + digits, rev + digits))


def encode(chars):
    return spaced(5, convert(chars))


def decode(chars):
    return ''.join(convert(chars))


def convert(chars):
    return filter(identity, (cipher[ch] for ch in chars.lower()))


def spaced(n, it):
    chunks = map(''.join, zip_longest(*([it] * n), fillvalue=''))
    return ''.join(chain(*zip(chunks, repeat(' ')))).rstrip()


identity = lambda x: x

File: exercism/javascript/atbash-cipher/atbash-cipher.js
const _ = { chunk, identity, fromPairs, zip } = require('lodash/fp')

const alpha = 'abcdefghijklmnopqrstuvwxyz'.split('')
const atbash = alpha.slice()
atbash.reverse()
const digits = '01234566789'.split('')
const conv = fromPairs(zip(alpha.concat(digits), atbash.concat(digits)))

const convert = (str) =>
  str.toLowerCase().split('').map(ch => conv[ch] || '').filter(identity)

const format = (str) =>
  chunk(5, str).map(chars => chars.join('')).join(' ')


const encode = (str) =>
  format(convert(str))

module.exports = { encode }
